In [1]:
# initial setup
%run "../../../common/0_notebooks_base_setup.py"


/Users/csuarezgurruchaga/Desktop/Digital-House/clase_11/clase_11/dsad_2021/common
default checking
Running command `conda list`... ok
jupyterlab=2.2.6 already installed
pandas=1.1.5 already installed
bokeh=2.2.3 already installed
seaborn=0.11.0 already installed
matplotlib=3.3.2 already installed
ipywidgets=7.5.1 already installed
pytest=6.2.1 already installed
chardet=4.0.0 already installed
psutil=5.7.2 already installed
scipy=1.5.2 already installed
statsmodels=0.12.1 already installed
scikit-learn=0.23.2 already installed
xlrd=2.0.1 already installed
Running command `conda install --yes nltk=3.5.0`... ok
Solving environment: ...working... done

# All requested packages already installed.


unidecode=1.1.1 already installed
pydotplus=2.0.2 already installed
pandas-datareader=0.9.0 already installed
flask=1.1.2 already installed


---

<img src='../../../common/logo_DH.png' align='left' width=35%/>


<a id="section_toc"></a> 
## Tabla de Contenidos

[Intro](#section_intro)

[Dataset](#section_dataset)

[Imports](#section_imports)

[Tidy data](#section_tidy_data)

[Repaso de joins con pandas](#section_repaso_joins)

[Referencias](#section_referencias)


---


# Pandas 2 - tidy data

<a id="section_intro"></a> 
## Intro

[volver a TOC](#section_toc)


Decimos que un dataset está ordenado cuando:
* Cada variable es una columna
* Cada observación es una fila
* Cada tipo de unidad observacional forma una tabla

Algunas definiciones:
* Variable: Es la medición de un atributo, por ejemplo, peso, altura, etc 
* Valor: Es la medida que toma una variable para una observación
* Observación: Todas las observaciones toman el mismo tipo de valores para cada variable. 



<a id="section_dataset"></a> 
## Dataset

[volver a TOC](#section_toc)


Usaremos el dataset de las obras del Met (The Metropolitan Museum of Art) que vimos en la clase de Limpieza de datos

https://github.com/metmuseum/openaccess/

Analizando el dataset del Met vemos que los datos de los artistas están en la misma tabla que los datos de las obras, incumpliendo la tercer regla que dice "Cada tipo de unidad observacional forma una tabla".

Veamos cómo resolver este punto.


<a id="section_imports"></a> 
## Imports

[volver a TOC](#section_toc)


In [2]:
import pandas as pd
import numpy as np

## Importamos los datos

Vemos de qué tamaño es el dataset e imprimimos losprimeros registros

In [3]:
# local
data_location = "../Data/MetObjects_sample.csv"

data = pd.read_csv(data_location)

data.dtypes

Unnamed: 0                   int64
Object Number               object
Is Highlight                  bool
Is Public Domain              bool
Is Timeline Work              bool
Object ID                    int64
Department                  object
AccessionYear               object
Object Name                 object
Title                       object
Culture                     object
Period                      object
Dynasty                     object
Reign                       object
Portfolio                   object
Artist Role                 object
Artist Prefix               object
Artist Display Name         object
Artist Display Bio          object
Artist Suffix               object
Artist Alpha Sort           object
Artist Nationality          object
Artist Begin Date           object
Artist End Date             object
Artist Gender               object
Artist ULAN URL             object
Artist Wikidata URL         object
Object Date                 object
Object Begin Date   

In [4]:
data.head(3)

Unnamed: 0  Object Number  Is Highlight  Is Public Domain  \
0      297513  Inst.1980.3.1         False             False   
1      366196     175T49 R43          True             False   
2      380983         WW.550         False             False   

   Is Timeline Work  Object ID           Department AccessionYear  \
0             False     442985          Islamic Art           NaN   
1             False     591826        The Libraries           NaN   
2             False     650203  Drawings and Prints           NaN   

               Object Name                    Title  ... Excavation River  \
0  Illustrated single work  Illustrated Single Work  ...        NaN   NaN   
1                      NaN      Vita del Tintoretto  ...        NaN   NaN   
2                    Print    Flowers and Dragonfly  ...        NaN   NaN   

  Classification Rights and Reproduction  \
0        Codices                     NaN   
1            NaN                     NaN   
2         Prints                     NaN   

                                       Link Resource  \
0  http://www.metmuseum.org/art/collection/search...   
1  http://www.metmuseum.org/art/collection/search...   
2  http://www.metmuseum.org/art/collection/search...   

                       Object Wikidata URL Metadata Date  \
0                                      NaN           NaN   
1  https://www.wikidata.org/wiki/Q29385980           NaN   
2                                      NaN           NaN   

                                 Repository                 Tags  \
0  Metropolitan Museum of Art, New York, NY          Abstraction   
1  Metropolitan Museum of Art, New York, NY                  NaN   
2  Metropolitan Museum of Art, New York, NY  Dragonflies|Flowers   

                                        Tags AAT URL  
0          http://vocab.getty.edu/page/aat/300056508  
1                                                NaN  
2  http://vocab.getty.edu/page/aat/300310476|http...  

[3 rows x 52 columns]

<a id="section_tidy_data"></a> 
## Tidy data

[volver a TOC](#section_toc)

Analizando el dataset del Met vemos que los datos de los artistas están en la misma tabla que los datos de las obras, incumpliendo la tercer regla que dice "Cada tipo de unidad observacional forma una tabla".

En este caso deberíamos tener una tabla para Obra y otra para Artista.

Separemos estos datos en dos tablas, y eliminemos duplicados en la tabla de artistas.

In [5]:
data_artist_columns = ['Artist Role', 'Artist Prefix', 'Artist Display Name',
       'Artist Display Bio', 'Artist Suffix', 'Artist Alpha Sort',
       'Artist Nationality', 'Artist Begin Date', 'Artist End Date',
       'Artist Gender', 'Artist ULAN URL', 'Artist Wikidata URL']

data_artist  = data.loc[:, data_artist_columns]
print(data_artist.shape)


(4743, 12)


In [6]:
data_artist.head(3)

Artist Role Artist Prefix             Artist Display Name  \
0            Artist           NaN            Mohammad Aslam Kamal   
1            Author           NaN                   Carlo Ridolfi   
2  Artist|Publisher           NaN  Maria Pranke|Wiener Werkstätte   

                                  Artist Display Bio Artist Suffix  \
0                                            b. 1939           NaN   
1                   Italian, Lonigo 1594–1658 Venice           NaN   
2  Hungarian, Pándorfalu Parndorf, Hungary 1891–1...           NaN   

                 Artist Alpha Sort Artist Nationality Artist Begin Date  \
0             Kamal Mohammad Aslam          Pakistani        1914         
1                   Ridolfi, Carlo            Italian        1594         
2  Pranke, Maria|Wiener Werkstätte          Hungarian        1891         

  Artist End Date Artist Gender                              Artist ULAN URL  \
0      1990                 NaN                                          NaN   
1      1658                 NaN   http://vocab.getty.edu/page/ulan/500013305   
2      1972                   |  |http://vocab.getty.edu/page/ulan/500268774   

                      Artist Wikidata URL  
0                                     NaN  
1   https://www.wikidata.org/wiki/Q776990  
2  |https://www.wikidata.org/wiki/Q241338

Veamos si hay registros duplicados en data_artist. 

Consideramos duplicados aquellos registros que coincidan en los campos "Artist Display Name", "Artist Role"

In [7]:
artist_duplicated = data_artist.duplicated(subset=["Artist Display Name", "Artist Role"])
any(artist_duplicated)

True

Eliminamos los registros duplicados:

In [8]:
data_artist_unique = data_artist.drop_duplicates(subset=["Artist Display Name", "Artist Role"], keep="first")
print(data_artist_unique.shape)
print(data_artist.shape)

(1615, 12)
(4743, 12)


Para no perder la asociación entre la obra y el artista, en la tabla data_object mantenemos los campos "Artist Display Name" y "Artist Role" que son los identificadores de artista en la tabla data_artist

In [9]:
data_artist_columns_sin_key =  ['Artist Prefix', 'Artist Display Bio', 'Artist Suffix', 'Artist Alpha Sort',
       'Artist Nationality', 'Artist Begin Date', 'Artist End Date',
       'Artist Gender', 'Artist ULAN URL', 'Artist Wikidata URL']
data_object = data.drop(data_artist_columns_sin_key, axis = 'columns')
print(data_object.shape)
data_object.columns

(4743, 42)


Index(['Unnamed: 0', 'Object Number', 'Is Highlight', 'Is Public Domain',
       'Is Timeline Work', 'Object ID', 'Department', 'AccessionYear',
       'Object Name', 'Title', 'Culture', 'Period', 'Dynasty', 'Reign',
       'Portfolio', 'Artist Role', 'Artist Display Name', 'Object Date',
       'Object Begin Date', 'Object End Date', 'Medium', 'Dimensions',
       'Credit Line', 'Geography Type', 'City', 'State', 'County', 'Country',
       'Region', 'Subregion', 'Locale', 'Locus', 'Excavation', 'River',
       'Classification', 'Rights and Reproduction', 'Link Resource',
       'Object Wikidata URL', 'Metadata Date', 'Repository', 'Tags',
       'Tags AAT URL'],
      dtype='object')

Cuando eliminamos los duplicados, la tabla de artistas pasó de tener 4743 registros a 1615.

Esto resulta en una mejora de performance en cuanto a espacio, y facilita mantener la consistencia de los valores en los registros. Como desventaja, vamos a necesitar combinar ambas tablas para poder responder algunas preguntas que involucen relaciones entre columnas en tablas distintas.



---
<a id="section_repaso_joins"></a> 

## Repaso de joins con pandas

[volver a TOC](#section_toc)

`merge` `concat` `join` `append`

Como ejercicio, hagamos ahora el camino inverso, a partir de dos tablas (objetos y artistas) combinemoslas para obtener un único DataFrame con el conjunto total de datos.

Como vimos en la práctica guiada 3, los datos contenidos en objetos pandas pueden combinarse usando los métodos

* `pandas.merge`: combina filas de dos DataFrames en base a una o más claves. Es análogo al join de SQL.

https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.merge.html
    
* `pandas.concat`: concatena o apila objetos pandas sobre alguno de los ejes

https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.concat.html

* `pandas.join`: agrega columnas de otro DataFrame

https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.join.html

* `pandas.append`: agrega al final las filas de otro DataFrame

https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.append.html

**Veamos un ejemplo de `merge`**, comparemos con su tamaño con el DataFrame original (antes de dividirlo en data_objet y data_artist)

In [10]:
data_all_merge = data_object.merge(data_artist_unique, left_on=['Artist Role', 'Artist Display Name'], 
                             right_on=['Artist Role', 'Artist Display Name'], how="left")

In [11]:
data_all_merge.shape

(4743, 52)

Estamos definiendo las columnas 'Artist Role', 'Artist Display Name' como claves de los registros para combinarlos en ambos DataFrames.

Dos filas van a ser combinadas en el DataFrame resultado si coinciden en los valores de las claves en los Dataset involucrados en el merge.

`left_on` indica qué columnas del primer DatFrame (el de la izquierda) son la clave en el merge

`right_on` indica qué columnas del segundo DatFrame (el de la derecha) son la clave en el merge

`how` indica como combinar los objetos
* inner: el resultado tiene sólo los registros cuyas claves están en ambos DataFrames
* left: el resultado tiene todos los registros del primer DataFrame (left) aunque la clave pueda no estar en el segundo DataFrame (right)
* rigth: el resultado tiene todos los registros del segundo DataFrame (right) aunque la clave pueda no estar en el primer DataFrame (left)



Veamos un ejemplo de `concat`, axis = 1 indica que concatena las columnas de los dos DataFrame. Observemos que no combina las columnas repetidas sino que el DataFrame resultado tiene una proveniente de cada DataFrame. 

In [13]:
data_all_concat = pd.concat([data_object, data_artist_unique],  axis = 1)
print(data_object.shape)
print(data_artist_unique.shape)
print(data_all_concat.shape)

data_all_concat.columns[data_all_concat.columns == 'Artist Display Name']

(4743, 42)
(1615, 12)
(4743, 54)


Index(['Artist Display Name', 'Artist Display Name'], dtype='object')

**Veamos un ejemplo de `concat`**, axis = 0 que indica que concatena las filas de los dos DataFrame. Observemos que no es necesario que el orden de las columnas sea el mismo en ambos DataFrames

In [14]:
data_object_1 = data_object.iloc[0:10, :]

columns_reverse = data_object.columns[::-1]
data_object_2 = data_object.iloc[10:20].loc[:, columns_reverse]

print(data_object_1.shape)
print(data_object_2.shape)

(10, 42)
(10, 42)


In [15]:
data_object_1_2 = pd.concat([data_object_1, data_object_2],  axis = 0)
print(data_object_1_2.shape)
data_object_1_2.sample(5)

(20, 42)


Unnamed: 0   Object Number  Is Highlight  Is Public Domain  \
10      411903         Ref.113         False             False   
7       257613          WW.229         False             False   
14      167662         X.684.1         False             False   
4       472119      24.63.2060         False             False   
12        6808  Inst.67.15.116         False             False   

    Is Timeline Work  Object ID           Department AccessionYear  \
10             False     708208  Drawings and Prints           NaN   
7              False     390830  Drawings and Prints           NaN   
14             False     271681          Photographs           NaN   
4              False     838838  Drawings and Prints           NaN   
12             False       7281    The American Wing           NaN   

   Object Name                                              Title  ...  \
10       Print  Plate 7: the story of the Pisan saints, here S...  ...   
7        Print                                     Fashion (Mode)  ...   
14  Photograph                  Salt Crystals on Lava, California  ...   
4        Print                Victor Hugo, published in Le Masque  ...   
12      Saucer                                             Saucer  ...   

   Excavation River Classification Rights and Reproduction  \
10        NaN   NaN         Prints                     NaN   
7         NaN   NaN         Prints                     NaN   
14        NaN   NaN    Photographs                     NaN   
4         NaN   NaN         Prints                     NaN   
12        NaN   NaN       Ceramics                     NaN   

                                        Link Resource Object Wikidata URL  \
10  http://www.metmuseum.org/art/collection/search...                 NaN   
7   http://www.metmuseum.org/art/collection/search...                 NaN   
14  http://www.metmuseum.org/art/collection/search...                 NaN   
4   http://www.metmuseum.org/art/collection/search...                 NaN   
12  http://www.metmuseum.org/art/collection/search...                 NaN   

   Metadata Date                                Repository           Tags  \
10           NaN  Metropolitan Museum of Art, New York, NY            NaN   
7            NaN  Metropolitan Museum of Art, New York, NY  Women|Fashion   
14           NaN  Metropolitan Museum of Art, New York, NY            NaN   
4            NaN  Metropolitan Museum of Art, New York, NY            NaN   
12           NaN  Metropolitan Museum of Art, New York, NY            NaN   

                                         Tags AAT URL  
10                                                NaN  
7   http://vocab.getty.edu/page/aat/300025943|http...  
14                                                NaN  
4                                                 NaN  
12                                                NaN  

[5 rows x 42 columns]

**Veamos un ejemplo de `join`**. Este método es muy eficiente (mejor que merge) porque usa los índices de los DataFrames. 

Para eso vamos a setear la columnas 'Artist Role' y 'Artist Display Name' como índice en los dos objetos.

Previamente debemos eliminar los registros que sean nulos en alguno de los dos campos de los DataFrames.

Los índices pueden ser creados sobre DataFrames con registros nulos en esos campos, pero el método join da error en esas condiciones.


In [16]:
print("cantidad de nulos en el campo 'Artist Role' de data_object_index", data_object['Artist Role'].isnull().sum())
print("cantidad de nulos en el campo 'Artist Role' de data_artist_unique_index", data_artist_unique['Artist Role'].isnull().sum())
print("cantidad de nulos en el campo 'Artist Display Name' de data_object_index", data_object['Artist Display Name'].isnull().sum())
print("cantidad de nulos en el campo 'Artist Display Name' de data_artist_unique_index", data_artist_unique['Artist Display Name'].isnull().sum())


cantidad de nulos en el campo 'Artist Role' de data_object_index 2359
cantidad de nulos en el campo 'Artist Role' de data_artist_unique_index 14
cantidad de nulos en el campo 'Artist Display Name' de data_object_index 2345
cantidad de nulos en el campo 'Artist Display Name' de data_artist_unique_index 1


In [17]:
data_artist_notnull = data_artist_unique.dropna(subset=['Artist Role', 'Artist Display Name'], how='any', axis=0)
data_object_notnull = data_object.dropna(subset=['Artist Role', 'Artist Display Name'], how='any', axis=0)

In [18]:
data_artist_notnull_index = data_artist_notnull.set_index(['Artist Role', 'Artist Display Name'])
data_object_notnull_index = data_object_notnull.set_index(['Artist Role', 'Artist Display Name'])

In [19]:
data_all_join = data_object_notnull_index.join(data_artist_notnull_index, lsuffix = "object_", rsuffix = "artist_")
print(data_all_join.shape)
data_all_join.sample(3)

(2384, 50)


Unnamed: 0  \
Artist Role                        Artist Display Name                                              
Designer|Design House              Karl Lagerfeld|Fendi                                     78782   
Artist|Artist                      Agostino Carracci|Francesco Francia                     209449   
Artist|Artist|Artist|Artist|Artist Kyūen Ryūchin|Kōshi Ehō|Nankō Sōgen|Zuigan Ryūs...       33834   

                                                                                      Object Number  \
Artist Role                        Artist Display Name                                                
Designer|Design House              Karl Lagerfeld|Fendi                                    2004.422   
Artist|Artist                      Agostino Carracci|Francesco Francia                 27.78.1(344)   
Artist|Artist|Artist|Artist|Artist Kyūen Ryūchin|Kōshi Ehō|Nankō Sōgen|Zuigan Ryūs...   1975.268.39   

                                                                                       Is Highlight  \
Artist Role                        Artist Display Name                                                
Designer|Design House              Karl Lagerfeld|Fendi                                       False   
Artist|Artist                      Agostino Carracci|Francesco Francia                        False   
Artist|Artist|Artist|Artist|Artist Kyūen Ryūchin|Kōshi Ehō|Nankō Sōgen|Zuigan Ryūs...         False   

                                                                                       Is Public Domain  \
Artist Role                        Artist Display Name                                                    
Designer|Design House              Karl Lagerfeld|Fendi                                           False   
Artist|Artist                      Agostino Carracci|Francesco Francia                            False   
Artist|Artist|Artist|Artist|Artist Kyūen Ryūchin|Kōshi Ehō|Nankō Sōgen|Zuigan Ryūs...              True   

                                                                                       Is Timeline Work  \
Artist Role                        Artist Display Name                                                    
Designer|Design House              Karl Lagerfeld|Fendi                                           False   
Artist|Artist                      Agostino Carracci|Francesco Francia                            False   
Artist|Artist|Artist|Artist|Artist Kyūen Ryūchin|Kōshi Ehō|Nankō Sōgen|Zuigan Ryūs...              True   

                                                                                       Object ID  \
Artist Role                        Artist Display Name                                             
Designer|Design House              Karl Lagerfeld|Fendi                                   107241   
Artist|Artist                      Agostino Carracci|Francesco Francia                    341094   
Artist|Artist|Artist|Artist|Artist Kyūen Ryūchin|Kōshi Ehō|Nankō Sōgen|Zuigan Ryūs...      39659   

                                                                                                Department  \
Artist Role                        Artist Display Name                                                       
Designer|Design House              Karl Lagerfeld|Fendi                                  Costume Institute   
Artist|Artist                      Agostino Carracci|Francesco Francia                 Drawings and Prints   
Artist|Artist|Artist|Artist|Artist Kyūen Ryūchin|Kōshi Ehō|Nankō Sōgen|Zuigan Ryūs...            Asian Art   

                                                                                      AccessionYear  \
Artist Role                        Artist Display Name                                                
Designer|Design House              Karl Lagerfeld|Fendi                                        2004   
Artist|Artist                      Agostino Carracci|Francesco Francia                         1927   
Artist|Artist

Para que el índice se transforme en dos columnas del DataFrame usamos el método `reset_index`

https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.reset_index.html

In [20]:
data_all_join = data_all_join.reset_index()
data_all_join.sample(3)

Artist Role  \
2292  Publisher|Artist|Artist   
1323         Artist|Publisher   
90                     Artist   

                                    Artist Display Name  Unnamed: 0  \
2292  Nicolas Le Cat|Pierre van Schuppen|Peter Paul ...      447781   
1323           Adalberta Kiessewetter|Wiener Werkstätte      380786   
90                     Anonymous, British, 19th century      466974   

             Object Number  Is Highlight  Is Public Domain  Is Timeline Work  \
2292           51.501.7177         False             False              True   
1323                WW.540         False             False             False   
90    Ref.BellaLandauer.54         False             False             False   

      Object ID           Department AccessionYear  ... Artist Prefix  \
2292     771092  Drawings and Prints          1951  ...         After   
1323     649938  Drawings and Prints           NaN  ...           NaN   
90       824091  Drawings and Prints           NaN  ...           NaN   

                                     Artist Display Bio Artist Suffix  \
2292  Flemish, active Antwerp, ca. 1650|Flemish, Ant...           NaN   
1323      Austrian, Leitmeritz, Bohemia 1887–after 1948           NaN   
90                                                  NaN           NaN   

                                      Artist Alpha Sort Artist Nationality  \
2292  Le Cat, Nicolas|Schuppen, Pierre van|Rubens, P...                NaN   
1323          Kiessewetter, Adalberta|Wiener Werkstätte           Austrian   
90                     Anonymous, British, 19th century                NaN   

                     Artist Begin Date                   Artist End Date  \
2292  1600      |1627      |1577        1700      |1702      |1640         
1323                        1887                                     NaN   
90                          1800                              1900         

     Artist Gender                                    Artist ULAN URL  \
2292            ||  |http://vocab.getty.edu/page/ulan/500032370|ht...   
1323             |        |http://vocab.getty.edu/page/ulan/500268774   
90             NaN                                                NaN   

                         Artist Wikidata URL  
2292   ||https://www.wikidata.org/wiki/Q5599  
1323  |https://www.wikidata.org/wiki/Q241338  
90                                       NaN  

[3 rows x 52 columns]

**Veamos un ejemplo de `append`**, que es similar a `concat`

In [21]:
data_object_1 = data_object.iloc[0:10, :]

columns_reverse = data_object.columns[::-1]
data_object_2 = data_object.iloc[10:20].loc[:, columns_reverse]

print(data_object_1.shape)
print(data_object_2.shape)

(10, 42)
(10, 42)


In [22]:
data_object_1_2 = data_object_1.append(data_object_2)
print(data_object_1_2.shape)
data_object_1_2.sample(5)

(20, 42)


Unnamed: 0  Object Number  Is Highlight  Is Public Domain  \
0       297513  Inst.1980.3.1         False             False   
8         5967  Inst.67.15.49         False             False   
18      416983    2020.64.125         False             False   
5       471850     2019.601.1         False             False   
3       379184         WW.211         False             False   

    Is Timeline Work  Object ID           Department AccessionYear  \
0              False     442985          Islamic Art           NaN   
8              False       6370    The American Wing           NaN   
18             False     715531    The American Wing           NaN   
5              False     838173    The American Wing           NaN   
3              False     647832  Drawings and Prints           NaN   

                Object Name  \
0   Illustrated single work   
8                     Plate   
18                     Vase   
5                   Drawing   
3                     Print   

                                                Title  ... Excavation River  \
0                             Illustrated Single Work  ...        NaN   NaN   
8                                               Plate  ...        NaN   NaN   
18                                               Vase  ...        NaN   NaN   
5                                          Four Flies  ...        NaN   NaN   
3   Old Karlsbad: City Tower (Alt-Karlsbad Stadtturm)  ...        NaN   NaN   

   Classification Rights and Reproduction  \
0         Codices                     NaN   
8        Ceramics                     NaN   
18       Ceramics                     NaN   
5        Drawings                     NaN   
3          Prints                     NaN   

                                        Link Resource Object Wikidata URL  \
0   http://www.metmuseum.org/art/collection/search...                 NaN   
8   http://www.metmuseum.org/art/collection/search...                 NaN   
18  http://www.metmuseum.org/art/collection/search...                 NaN   
5   http://www.metmuseum.org/art/collection/search...                 NaN   
3   http://www.metmuseum.org/art/collection/search...                 NaN   

   Metadata Date                                Repository         Tags  \
0            NaN  Metropolitan Museum of Art, New York, NY  Abstraction   
8            NaN  Metropolitan Museum of Art, New York, NY          NaN   
18           NaN  Metropolitan Museum of Art, New York, NY          NaN   
5            NaN  Metropolitan Museum of Art, New York, NY          NaN   
3            NaN  Metropolitan Museum of Art, New York, NY       Towers   

                                 Tags AAT URL  
0   http://vocab.getty.edu/page/aat/300056508  
8                                         NaN  
18                                        NaN  
5                                         NaN  
3   http://vocab.getty.edu/page/aat/300004847  

[5 rows x 42 columns]

<a id="section_referencias"></a> 
## Referencias

[volver a TOC](#section_toc)


Python for Data Analysis. Wes McKinney. Cap 8.2

https://pandas.pydata.org/pandas-docs/stable/user_guide/merging.html